### PR TITLE
[RTL] Add result type verification of instance op

### LINKED ISSUE #617: [RTL] Instance should verify that its return types match the Module
RTL does not appear to be checking that the number of arguments returned by InstanceOp match the module it is referencing.  The following does not produce an error:

```mlir
// RUN: circt-opt ./test.mlir

rtl.module @empty() -> () {
  rtl.output
}

rtl.module @test() -> () {
  // too many outputs
  %0, %1, %3 = rtl.instance "inline-me" @mem() : () -> (i2, i2, i2)
  rtl.output
}
```

### PR DESCRIPTION
**Pull request by uenoku**
This PR adds the following verification for instance op result types. 

* The numbers of results are the same
* Result types are the same (except for none type)

Will close #617

### CODE REVIEW

**Comment by youngar** [comment]
Thanks!  It doesn't look like there is any parameter verification either, is that also missing?

**Comment by teqdruid** [comment]
There's problem with type checking of extern modules -- we support parameterized external modules, but don't have proper parameterization support in the IR. As a result, we often intentionally, necessarily mismatch operand and result types. I don't know of cases where we need to specify the wrong number of operands and/or results.

So this needs to be confined to modules which are defined in the IR (ignore extern modules).

**Comment by uenoku** [comment]
> It doesn't look like there is operand verification either, is that also missing?
Yes, it's missing. I'll add. 

> So this needs to be confined to modules which are defined in the IR (ignore extern modules).
OK

**Comment by youngar** [comment]
@teqdruid Thanks for pointing that out!  I finally understand what that `parameters` attribute is for :). Would it make sense to skip type checking only if the `rtl.instance` has a `parameters` attribute?  Would it also make sense to ensure that the `parameter` attribute is only there for external modules?

File: lib/Dialect/RTL/RTLOps.cpp [hunk 1]
```diff
@@ -459,6 +459,35 @@ static LogicalResult verifyInstanceOp(InstanceOp op) {
...
+    auto diag = op.emitOpError()
+                << "has a wrong number of results; expected "
+                << expectedTypes.size() << " but got " << numResults;
+    diag.attachNote(referencedModule->getLoc())
+        << "original module declared here";
+
+    return failure();
+  }
+
+  for (size_t i = 0; i != numResults; i++) {
```
Comment (lattner) [inline lib/Dialect/RTL/RTLOps.cpp:476]: Super nit, prefer preincrement `++i` for consistency.

**Review by lattner — APPROVED** [review]
awesome, thank you for tackling this.  Out of curiosity, why the special handling for NoneType?  Can you add a testcase for it if we need it?

**Comment by teqdruid** [comment]
I don’t think that disabling the check if there are parameters is sufficient as an external module can have default parameter values which affect the types. Only doing the check for rtl.module seems reasonable to me.

Also, could you run the integration tests before merging?

thanks!

**Comment by uenoku** [comment]
> Out of curiosity, why the special handling for NoneType? Can you add a testcase for it if we need it?
I added the special handling because I encountered the test failure for test/Dialect/ESI/lowering.mlir. 

It looks none type is used for parameterization for the external module so there is a mismatch of type.
For example,
```
 rtl.module.extern @ESI_PipelineStage(i1 {rtl.name = "clk"}, i1 {rtl.name = "rstn"}, none {rtl.name = "a"}, i1 {rtl.name = "a_valid"}, i1 {rtl.name = "x_ready"}) -> (%a_ready: i1, %x: none, %x_valid: i1)
...
 %pipelineStage.a_ready, %pipelineStage.x, %pipelineStage.x_valid = rtl.instance "pipelineStage" @ESI_PipelineStage(%clk, %rstn, %3, %2, %pipelineStage.a_ready_0) {parameters = {WIDTH = 4 : ui32}} : (i1, i1, i4, i1, i1) -> (i1, i4, i1)
```

But as @teqdruid suggested, I'll remove the verification for the external module. Therefore, I guess I can remove the handling of none type too.

File: lib/Dialect/RTL/RTLOps.cpp (hunk 1 above)
Comment (uenoku) [reply lib/Dialect/RTL/RTLOps.cpp]: addressed

File: lib/Dialect/RTL/RTLOps.cpp [hunk 2]
```diff
@@ -459,6 +459,66 @@ static LogicalResult verifyInstanceOp(InstanceOp op) {
       return failure();
   }
 
+  // If the referenced moudle is internal, check that input and result types are
+  // consistent with the referenced module.
+  if (isa<RTLModuleOp>(referencedModule)) {
```
Comment (teqdruid) [inline lib/Dialect/RTL/RTLOps.cpp:464]: As per the llvm style guide prefer early exits. e.g. if not rtlmoduleop, return.

File: lib/Dialect/RTL/RTLOps.cpp [hunk 3]
```diff
@@ -459,6 +459,66 @@ static LogicalResult verifyInstanceOp(InstanceOp op) {
       return failure();
   }
 
+  // If the referenced moudle is internal, check that input and result types are
```
Comment (teqdruid) [inline lib/Dialect/RTL/RTLOps.cpp:462]: nit: the verify function is a bit long for my taste, so I would break out the new code into another function.

**Review by teqdruid — APPROVED** [review]
Thanks! Great error messages!

File: lib/Dialect/RTL/RTLOps.cpp (hunk 3 above)
Comment (teqdruid) [reply lib/Dialect/RTL/RTLOps.cpp:462]: just personal preference so feel free to ignore.

Comment (uenoku) [reply lib/Dialect/RTL/RTLOps.cpp:526]: I have split the function

**Comment by lattner** [comment]
> But as @teqdruid suggested, I'll remove the verification for the external module. Therefore, I guess I can remove the handling of none type too.

Sounds good, I think we need to continue working on the representation of external modules.  This is great progress for normal ones!

**Review by lattner — APPROVED** [review]
Patch looks great, thanks again!

**Pull request merged**